webapp/src/ts/modules/contacts/contacts-edit.component.ts

```diff
@@ -145,24 +145,46 @@ export class ContactsEditComponent implements OnInit, OnDestroy, AfterViewInit {
     this.initForm();
   }
 
-  private initForm() {
+  private async initForm() {
     this.contentError = false;
     this.errorTranslationKey = false;
 
-    return this
-      .getContact()
-      .then((contact) => this.getForm(contact))
-      .then((formId) => this.renderForm(formId))
-      .then((formInstance) => this.setEnketoContact(formInstance))
-      .then(() => {
-        this.globalActions.setLoadingContent(false);
-      })
-      .catch((err) => {
-        this.errorTranslationKey = err.translationKey || 'error.loading.form';
-        this.globalActions.setLoadingContent(false);
-        this.contentError = true;
-        console.error('Error loading contact form.', err);
-      });
+    try {
+      const contact = await this.getContact();
+      const contactTypeId = this.contactTypesService.getTypeId(contact) || this.routeSnapshot.params?.type;
+      const contactType = await this.contactTypesService.get(contactTypeId);
+      if (!contactType) {
+        throw new Error(`Unknown contact type "${contactTypeId}"`);
+      }
+
+      const titleKey = contact ? contactType.edit_key : contactType.create_key;
+      const formId = this.getForm(contact, contactType, titleKey);
+      if (!formId) {
+        throw new Error('Unknown form');
+      }
+
+      const formDoc = await this.dbService.get().get(formId);
+      const instanceData = this.getFormInstanceData();
+      const markFormEdited = this.markFormEdited.bind(this);
+      const resetFormError = this.resetFormError.bind(this);
+      const formContext: EnketoFormContext = {
+        selector: '#contact-form',
+        formDoc,
+        instanceData,
+        editedListener: markFormEdited,
+        valuechangeListener: resetFormError,
+        titleKey,
+      };
+      const formInstance = await this.renderForm(formContext);
+      this.setEnketoContact(formInstance);
```

Comment:
I think we can slim this function down like: 
```suggestion
      const formId = this.getForm(contact, contactType);
      if (!formId) {
        throw new Error('Unknown form');
      }

      const titleKey = contact ? contactType.edit_key : contactType.create_key;
      this.setTitle(titleKey);
      const formInstance = await this.renderForm(formId, titleKey);
      this.setEnketoContact(formInstance);
```